core/src/main/java/com/linecorp/armeria/internal/http/Http1ObjectEncoder.java

```diff
@@ -300,60 +302,72 @@ protected ChannelFuture doWriteData(
         }
 
         try {
-            if (isTls && length > MAX_TLS_DATA_LENGTH) {
+            if (!isTls || length <= MAX_TLS_DATA_LENGTH) {
+                // Cleartext connection or data.length() <= MAX_TLS_DATA_LENGTH
+                return doWriteUnsplitData(ctx, id, data, endStream);
+            } else {
                 // TLS and data.length() > MAX_TLS_DATA_LENGTH
                 return doWriteSplitData(ctx, id, data, endStream);
-            } else {
-                // Cleartext connection or data.length() <= MAX_TLS_DATA_LENGTH
-                final ByteBuf buf = toByteBuf(ctx, data);
-                final HttpContent content;
-                if (endStream) {
-                    content = new DefaultLastHttpContent(buf);
-                } else {
-                    content = new DefaultHttpContent(buf);
-                }
-                ChannelFuture future = write(ctx, id, content, endStream);
-                ctx.flush();
-                return future;
             }
         } catch (Throwable t) {
             return ctx.newFailedFuture(t);
         }
     }
 
+    private ChannelFuture doWriteUnsplitData(ChannelHandlerContext ctx, int id, HttpData data,
+                                             boolean endStream) {
+        final ByteBuf buf = toByteBuf(ctx, data);
+        boolean written = false;
+        try {
+            final HttpContent content;
+            if (endStream) {
+                content = new DefaultLastHttpContent(buf);
+            } else {
+                content = new DefaultHttpContent(buf);
+            }
+            ChannelFuture future = write(ctx, id, content, endStream);
+            written = true;
```

Comment:
`Http1ObjectEncoder.write()` may return `ctx.newFailedFuture(ClosedPublisherException.get())`. Is this okay that we treat a failed future as `written`?